addons/binding/org.openhab.binding.energenie/src/main/java/org/openhab/binding/energenie/handler/EnergenieGatewayHandler.java

```diff
@@ -0,0 +1,619 @@
+    private boolean isGatewayThingAlreadyRegistered(Thing thing) {
+        ServiceReference ref = bundleContext.getServiceReference(ThingRegistry.class.getName());
+        thingRegistry = (ThingRegistry) bundleContext.getService(ref);
+        Collection<Thing> things = thingRegistry.getAll();
+        for (Thing registeredThing : things) {
+            ThingUID registeredThingUID = registeredThing.getUID();
+            ThingTypeUID registeredThingTypeUID = registeredThing.getThingTypeUID();
+            if (registeredThingTypeUID != null
+                    && registeredThingTypeUID.equals(EnergenieBindingConstants.THING_TYPE_GATEWAY)) {
+                Configuration config = registeredThing.getConfiguration();
+                String code = (String) config.get(EnergenieBindingConstants.CONFIG_GATEWAY_CODE);
+                if (gatewayCode != null && gatewayCode.equals(code) && (!thing.getUID().equals(registeredThingUID))) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    public void registerNewGateway() {
+        JsonObject registrationResponse = apiManager.registerGateway(getThing().getLabel(), gatewayCode);
+        if (registrationResponse != null) {
+            JsonObject responseData = registrationResponse.get(JSONResponseConstants.DATA_KEY).getAsJsonObject();
+            gatewayId = responseData.get(DeviceConstants.DEVICE_ID_KEY).getAsInt();
+            setProperties(responseData);
+        } else {
+            logger.warn(NOT_SUCCESSFULL_REGISTRATION_MESSAGE);
+        }
+    }
+
+    private void scheduleUpdateTask(long updateInterval) {
+        if (updateTaskResult != null) {
+            this.updateTaskResult.cancel(true);
+            this.updateTaskResult = null;
+        }
+        /**
+         * The refresh thread which is used for checking the "last_seen_at"
+         * property of the gateway
+         */
+        Runnable refreshRunnable = () -> {
+            logger.debug("Starting the refresh thread...");
+            checkGateway();
+        };
+        try {
+            scheduler.scheduleWithFixedDelay(refreshRunnable, INITIAL_REFRESH_DELAY_SECONDS, updateInterval,
+                    TimeUnit.SECONDS);
+        } catch (RejectedExecutionException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    REJECTED_EXECUTION_OF_UPDATE_TASK_MESSAGE);
+        }
+    }
+
+    /**
+     * Checks when the gateway was last active and updates the thing's channel.
+     * This method is called periodically by a refresh thread.
+     */
+    public void checkGateway() {
+        /*
+         * Mi|Home REST API does not provide a method for showing a gateway's
+         * information. So in order to check when was the last time it was seen
+         * you need to list all gateways and find the one you want by its id.
+         */
+        JsonObject gateway = getGatewayById(gatewayId);
```

Comment:
Did you really create mapper objects in your unit tests and use `JsonObject`s here, pitty.